File: sample_node.cpp
// A sample code for assignment #1 using C++ class

#include <ros/ros.h>
#include <iostream>     			// to use: std::cout, std::cin, etc
#include <cmath>					// to use: atan2, sqrt, pow, etc
#include <iomanip>     				// to use: setprecision()
#include <geometry_msgs/Twist.h> 
#include <turtlesim/Pose.h>  


using namespace std;

/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/
// THIS IS THE TURTLE CLASS

class TurtleClass
{
	public:
		/*/////////////////////////////////////////////////////////////*/
		/*/////////////////////////////////////////////////////////////*/			
		// constructor prototype
		TurtleClass(ros::NodeHandle& n);

		/*/////////////////////////////////////////////////////////////*/
		/*/////////////////////////////////////////////////////////////*/			
		// DESTRUCTOR
		~TurtleClass()
		{
		}

	/*/////////////////////////////////////////////////////////////*/
	/*/////////////////////////////////////////////////////////////*/	
	// THESE ARE THE PRIVATE VARIABLES
	private:
        ros::Subscriber sub;				// create subscriber object;
        ros::Publisher pub;					// create publisher object;
		turtlesim::Pose cp;					// current position
		turtlesim::Pose dp;					// desired position	
		turtlesim::Pose err;				// error (Pose: x,y,z rot_x, rot_y,rot_z)
		int counter;						// counter
		static const float tol = 0.1;		// error tolerance
        static const float p_gain = 0.5;	// proportional gain (default: 0.5)
        float err_euc;                      // error (Euclidean distance between current and goal position)	

		// Functions prototypes...
		void TurtleCallback(const turtlesim::Pose& msg);
		void TurtleGetVal(void);
		void TurtleError(void);
		void TurtleMove(void);

};
/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/
// CONSTRUCTOR
TurtleClass::TurtleClass(ros::NodeHandle& n)
{
    // create publisher object, will be used to send to turtle1/cmd_vel topic with queue size of 10
    pub = n.advertise<geometry_msgs::Twist>("turtle1/cmd_vel",10);

    // create subscriber object based on TurtleClass::TurtleCallback function
	sub = n.subscribe("turtle1/pose",10,&TurtleClass::TurtleCallback,this); 

	// define loop rate to be 10Hz
	ros::Rate loop_rate(10);
	counter = 0;
	// runs this loop until there is terminate singal
	while(ros::ok()) 
	{	
		counter++;          // increase counter
		ros::spinOnce();	// call callback functions

		// delay in running callback functions (roughly taks 1-2 counters)
		if (counter > 3)
		{
			// compute error
			TurtleError();

			// when error to the goal becomes small enough --OR-- if this is the 1st time, 
			if (abs(err_euc) < tol || counter == 4)
			{
		        // display some messeages to screen and asks user for the next inputs...

				TurtleGetVal();
			}
			// otherwise keep moving the turtle
			else
			{
				TurtleMove();
			}
		}
		// this is necessary to prevent excess cpu usage (this is while loop)
		loop_rate.sleep();
	}																																		
}

/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/			
// CALLBACK FUNCTION
void TurtleClass::TurtleCallback(const turtlesim::Pose& msg)
{
	// setting current position/orientation values with the messages received 
	// from the turtlesim/Pose topic
    cp.x = msg.x;               // current position (x)
    cp.y = msg.y;               // current position (y)
    cp.theta = msg.theta;       // current orientation (theta)
}

/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/	
// DISPLAY POSE, GET USER INPUT
void TurtleClass::TurtleGetVal(void)
{
    /*/////////////////////////////////////////////////////////////*/
    /*/////////////////////////////////////////////////////////////*/
    // display current status (Pose, distance to goal) of the turtle

	cout << endl;
	cout << "---------------------------------------------" << endl;
	cout << "Current pose [xc,yc,th]=[" << setprecision(3) << cp.x << "," 
		<< setprecision(3) << cp.y << "," << setprecision(3) << cp.theta << "]" << endl;
	cout << "Previous goal [xg,yg]=[" << setprecision(3) << dp.x << "," 
		<< setprecision(3) << dp.y << "]" << endl;
	cout << "Error ||xg-xc,yg-yc||=" << setprecision(3) << err_euc 
		<< " (err tol=" << tol << ")" << endl;
	cout << "---------------------------------------------" << endl;			
	cout << endl;	
	cout << "Enter x y coordindates of the goal (e.g., 10 10 means x=10,y=10)" << endl;
	cout << "or CTRL+C to exit" << endl;
    cout << endl;

    /*/////////////////////////////////////////////////////////////*/
    /*/////////////////////////////////////////////////////////////*/


	// get user input...
    cin >> dp.x >> dp.y;                // desires position (x y)
}

/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/	
// COMPUTE POSE ERRORS
void TurtleClass::TurtleError(void)
{
    // correct way to compute angle (using atan2(y,x))
	dp.theta = atan2(dp.y-cp.y,dp.x-cp.x);
	
	// angle difference between turtle's orientation and angle to goal position 				
	err.theta = atan2(sin(dp.theta - cp.theta),cos(dp.theta - cp.theta));

	// position error between current and goal position
    err.x = dp.x - cp.x;            // error in x
    err.y = dp.y - cp.y;            // error in y
	// euclidean distance from current to goal position
    err_euc =  sqrt(pow(err.x,2) + pow(err.y,2));	// error sqrt(x^2 + y^2)
}

/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/	
// GENERATE CONTROL, AND SEND PUBLISH IT TO TURTLE
void TurtleClass::TurtleMove(void)
{
    geometry_msgs::Twist vel;               // create publisher object vel with messeage type geometry_msgs::Twist

    // proportional controller
    vel.linear.x = p_gain * err_euc;        // linear velocity (x)
    vel.angular.z = p_gain * err.theta;     // angular velocity (z)
	
    // publish data	to 'vel'
	pub.publish(vel);
	
	// display distance to goal...
	cout << left << setw(20) << "Turtle is moving! " << setw(15) << "Distance to goal=" 
		<< setw(20) << setprecision(3) << err_euc << endl;
}	


/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/
/*/////////////////////////////////////////////////////////////*/
// MAIN FUNCTION

int main(int argc, char **argv)
{
	// create node 'sample_node' (ros::init_options::NoSigintHandler works file for no reason, 
	// I need to figure out, without it if you try Ctrl+C at cin, it won't work...) 
	ros::init(argc, argv,"sample_node",ros::init_options::NoSigintHandler);  
    ros::NodeHandle n;					// create node handle
    // create a TurtleClass
	TurtleClass turtle(n);

	return 0;
}

File: sample_node.py
#!/usr/bin/env python
import rospy
import math
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
import numpy as np

class TurtleClass():
	def __init__(self):
		cnt = 0
		tol = 0.1
		tol_ang = 0.01
		err_x = 0.0
		err_y = 0.0
		self.err_euc = 0.0
		self.err_theta = 0.0
		self.err_rot = 0.0
		self.vel=Twist()
		self.c_pos = Pose()
		d_pos = Pose()
		pub = rospy.Publisher('turtle1/cmd_vel', Twist, queue_size=10)
		sub = rospy.Subscriber('turtle1/pose', Pose, self.PoseCallback)
		rospy.init_node('sample_node')
	 	loop_rate = rospy.Rate(10)
		while not rospy.is_shutdown():
			if cnt > 0:
				d_pos_th = math.atan2(float(d_pos.y)-self.c_pos.y,float(d_pos.x)-self.c_pos.x)
				self.err_theta = math.atan2(math.sin(d_pos_th - self.c_pos.theta),math.cos(d_pos_th - self.c_pos.theta));
				err_x = float(d_pos.x) - self.c_pos.x
				err_y = float(d_pos.y) - self.c_pos.y
				self.err_euc =  math.sqrt(math.pow(err_x,2) + math.pow(err_y,2))
				ang = math.fmod(float(d_pos.theta),2*math.pi) - math.fmod(self.c_pos.theta, (2*math.pi))
				if abs(ang) > math.pi:
					self.err_rot = -np.sign(ang) * (math.pi*2 - abs(ang))
				else:
					self.err_rot = ang
			cnt+=1
				
			if abs(self.err_euc) < tol and abs(self.err_rot) < tol_ang:
				disp_ang = math.fmod(self.c_pos.theta,2*math.pi)
				if abs(disp_ang) > math.pi:
					disp_ang = -np.sign(disp_ang) * (2*math.pi  - abs(disp_ang))
				print "---------------------------------------------"
				print "Current pose [xc,yc,thc]=[%4.2f,%4.2f,%4.2f]" %(self.c_pos.x,self.c_pos.y,disp_ang)
				print "Previous goal [xg,yg,thg]=[%4.2f,%4.2f,%4.2f]" %(float(d_pos.x),float(d_pos.y),math.fmod(float(d_pos.theta),2*math.pi))
				print "Error, position:||xg-xc,yg-yc||=%4.2f, orientation:||thg-thc||=%4.2f" % (self.err_euc,self.err_rot)
				print "---------------------------------------------\n"
				d_pos.x, d_pos.y, d_pos.theta = raw_input("Enter desired goal pose (e.g., 10 10 2 means xg=10,yg=10,thg=2)\n").split()
			elif abs(self.err_euc) >= tol:
				self.move()
				#print "%4.2f, %4.2f" %(self.vel.linear.x, self.vel.angular.z)
				pub.publish(self.vel)
				print "Turtle is moving!, distance to goal=%f" % (self.err_euc) 
			else:
				self.rotate()
				print "Turtle is rotating!, orientation error=%f" % (self.err_rot)
				pub.publish(self.vel)
			loop_rate.sleep()														

			
		
	def PoseCallback(self, msg):
		#global c_pos
		self.c_pos.x = msg.x
		self.c_pos.y = msg.y
		self.c_pos.theta = msg.theta

	def move(self):
		p_gain = 0.5
		self.vel.linear.x = p_gain * self.err_euc
		self.vel.angular.z = p_gain * self.err_theta

	def rotate(self):
		p_gain = 0.5
		self.vel.linear.x = 0
		self.vel.angular.z = p_gain * self.err_rot

if __name__ == '__main__':

	

	print "+===========================================+" 
	print "|                                           |"
	print "| A sample program for ECE550 assignment #1 |"
	print "|                                           |"
	print "| by Hyongju Park (1/27/2016)               |"
	print "|                                           |"
	print "+===========================================+\n"

	try:
		turtle = TurtleClass()
	except rospy.ROSInterruptException: 
		pass



